class library: only schedule messages if necessary

telephon opened the pull request:
The message schedBundleArrayOnClock is sent to a number in order to
send off an OSC bundle a after certain time. If that time equals zero,
there is no need to add the task to the clock’s queue. This commit adds
a case where the bundle is send right now without scheduling it on the
clock in that case.

This fixes #3533.

We should discuss the possible implication for the case that a specific clock is expected as the context of scheduling.

telephon commented:
Here are some tests. Surprisingly, the times are not identical as they should be for *both* branches.

```supercollider

// make responders
(
var t, f;
f = { |msg, time| 
	[msg, time].postln; 
	if(t.notNil) { 
		"time difference:% ms".format(absdif(t, time) * 1000).postln;
		t = nil;
	} { 
		t = time
	}; 
};

OSCFunc(f, "/pong1");
OSCFunc(f, "/pong2");

s.waitForBoot {
	SynthDef(\ping1, { SendReply.ar(Impulse.ar(0), '/pong1'); FreeSelf.kr(Impulse.kr(0)) }).add;
	SynthDef(\ping2, { SendReply.ar(Impulse.ar(0), '/pong2'); FreeSelf.kr(Impulse.kr(0)) }).add;
}
)

// not at the same time
(
TempoClock.default.sched(0, { s.sendBundle(0.5, ["/s_new", "ping1", -1, 1, 1]) });
0.2.schedBundleArrayOnClock(TempoClock.default, [["/s_new", "ping2", -1, 1, 1]], 0, s, 0.3);

)

// not at the same time
(
0.0.schedBundleArrayOnClock(TempoClock.default, [["/s_new", "ping1", -1, 1, 1]], 0, s, 0.5);
0.2.schedBundleArrayOnClock(TempoClock.default, [["/s_new", "ping2", -1, 1, 1]], 0, s, 0.3);
)

// not at the same time
(
fork {
	0.0.schedBundleArrayOnClock(TempoClock.default, [["/s_new", "ping1", -1, 1, 1]], 0, s, 0.5);
	0.2.schedBundleArrayOnClock(TempoClock.default, [["/s_new", "ping2", -1, 1, 1]], 0, s, 0.3);
}
)
```

telephon commented:
But maybe this is the `time` in the OSC responder?

brianlheim commented on SCClassLibrary/Common/Math/SimpleNumber.sc:
```diff
@@ -523,14 +523,21 @@ SimpleNumber : Number {
+				server.sendBundle(latency ? server.latency, *bundleArray)
+			}
+		}
+
```
if you rewrote this with the scheduling bit as a lambda, I think it would be a lot easier to read:

```supercollider
var schedFunc = { server.sendBundle(latency ? server.latency, *bundleArray) };
if(lag != 0) {
  clock.sched(this, { SystemClock.sched(lag, schedFunc) })
} {
  if(this != 0) {
    clock.sched(this, schedFunc)
  } {
    schedFunc.value
  }
}
```

brianlheim replied:
A comment explaining the different cases couldn't hurt, either

telephon commented:
I've added an extra case (delta = 0, lag != 0), and refactored as brian suggested.
@brianlheim @adcxyz @joshpar looks good?

I'd love to refactor this from ground up, but there are some constraints on efficiency (passing args is a bottleneck) and extendibility (in event).
